src/transformers/cache_utils.py

```diff
@@ -146,44 +185,35 @@ def update(
                 The new value states to cache.
             layer_idx (`int`):
                 The index of the layer to cache the states for.
-            cache_kwargs (`dict[str, Any]`, `optional`):
+            cache_kwargs (`dict[str, Any]`, *optional*):
                 Additional arguments for the cache subclass. These are specific to each subclass and allow new types of
                 cache to be created.
 
         Return:
             A tuple containing the updated key and value states.
         """
-        raise NotImplementedError("Make sure to implement `update` in a subclass.")
-
-    def get_seq_length(self, layer_idx: Optional[int] = 0) -> int:
-        """Returns the sequence length of the cached states. A layer index can be optionally passed."""
-        # TODO: deprecate this function in favor of `cache_position`
-        raise NotImplementedError("Make sure to implement `get_seq_length` in a subclass.")
+        self.append_new_layers(layer_idx)
+        self.key_cache[layer_idx], self.value_cache[layer_idx] = self.layers[layer_idx].update(
+            key_states, value_states, self.key_cache[layer_idx], self.value_cache[layer_idx], cache_kwargs
+        )
+        return self.key_cache[layer_idx], self.value_cache[layer_idx]
```

Comment:
See comment above - each layer should be fully able to return whatever it wants, and save whatever it wants that may be different from its content. This is only possible if the layer hold their own data (see suggestion for what it should look like).
As of now, I believe this update is wrong when layers need to return something else than what they cache, i.e. a sliding update with prefill larger than its sliding window!! In those case, we cache less than what we return, which is not the case here
```suggestion
        self.append_new_layers(layer_idx)
        return self.layers[layer_idx].update(
            key_states, value_states, , cache_kwargs
        )
```